pkg/cloudprovider/openstack/openstack.go

```diff
@@ -625,28 +626,41 @@ func (lb *LoadBalancer) UpdateTCPLoadBalancer(name, region string, hosts []strin
 func (lb *LoadBalancer) DeleteTCPLoadBalancer(name, region string) error {
 	glog.V(4).Infof("DeleteTCPLoadBalancer(%v, %v)", name, region)
 
-	vip, err := getVipByName(lb.network, name)
-	if err != nil {
-		return err
+	vip, vipErr := getVipByName(lb.network, name)
+	vipExists := true
+	if vipErr != nil {
+		if vipErr != ErrNotFound {
+			return vipErr
+		}
+		vipExists = false
 	}
 
-	pool, err := pools.Get(lb.network, vip.PoolID).Extract()
-	if err != nil {
-		return err
+	pool, poolErr := pools.Get(lb.network, vip.PoolID).Extract()
+	poolExists := true
+	if poolErr != nil {
+		detailedErr, ok := poolErr.(*gophercloud.UnexpectedResponseCodeError)
+		if !ok || detailedErr.Actual != http.StatusNotFound {
+			return poolErr
+		}
+		poolExists = false
 	}
 
 	// Have to delete VIP before pool can be deleted
-	err = vips.Delete(lb.network, vip.ID).ExtractErr()
-	if err != nil {
-		return err
+	if vipExists {
+		err := vips.Delete(lb.network, vip.ID).ExtractErr()
+		if err != nil {
+			return err
+		}
 	}
 
 	// Ignore errors for everything following here
 
-	for _, monId := range pool.MonitorIDs {
-		pools.DisassociateMonitor(lb.network, pool.ID, monId)
+	if poolExists {
+		for _, monId := range pool.MonitorIDs {
+			pools.DisassociateMonitor(lb.network, pool.ID, monId)
+		}
+		pools.Delete(lb.network, pool.ID)
```

Comment:
I'd think so as well, but I don't know why @anguslees originally wrote it this way, so I'm hesitant to change it, given that it looks to have been an intentional choice. It'd be nice if he could chime in, but otherwise I don't want to mess with it given that I have no way of testing it (since I don't have an openstack cluster).